plaso/parsers/postgresql.py

```diff
@@ -0,0 +1,214 @@
+    """Builds time elements from an APT History time stamp.
+
+    Args:
+      time_elements_structure (pyparsing.ParseResults): structure of tokens
+          derived from an APT History time stamp.
+
+    Returns:
+      dfdatetime.TimeElements: date and time extracted from the structure or
+          None if the structure does not represent a valid string.
+    """
+    # Ensure time_elements_tuple is not a pyparsing.ParseResults otherwise
+    # copy.deepcopy() of the dfDateTime object will fail on Python 3.8 with:
+    # "TypeError: 'str' object is not callable" due to pyparsing.ParseResults
+    # overriding __getattr__ with a function that returns an empty string when
+    # named token does not exist.
+    try:
+      if len(time_elements_structure) == 6:
+        year, month, day_of_month, hours, minutes, seconds = (
+            time_elements_structure)
+
+        date_time = dfdatetime_time_elements.TimeElements(time_elements_tuple=(
+            year, month, day_of_month, hours, minutes, seconds))
+
+      else:
+        year, month, day_of_month, hours, minutes, seconds, milliseconds = (
+            time_elements_structure)
+
+        date_time = dfdatetime_time_elements.TimeElementsInMilliseconds(
+            time_elements_tuple=(
+                year, month, day_of_month, hours, minutes, seconds,
+                milliseconds))
+
+      date_time.is_local_time = True
+      return date_time
+    except (TypeError, ValueError):
+      return None
+
+  def ParseRecord(self, parser_mediator, key, structure):
+    """Parses a record and produces a PostgreSQL event.
+
+    Args:
+      parser_mediator (ParserMediator): mediates interactions between parsers
+          and other components, such as storage and dfVFS.
+      key (str): name of the parsed structure.
+      structure (pyparsing.ParseResults): elements parsed from the file.
+
+    Raises:
+      ParseError: when the structure type is unknown.
+    """
+    if key not in self._SUPPORTED_KEYS:
+      raise errors.ParseError(
+          'Unable to parse record, unknown structure: {0:s}'.format(key))
+
+    event_data = PostgreSQLEventData()
+    event_data.pid = ''.join(
+        [str(pid) for pid in self._GetValueFromStructure(structure, 'pid')])
+
+    log_level = self._GetValueFromStructure(structure, 'log_level')
+    if log_level and len(log_level) != 1:
+      parser_mediator.ProduceExtractionWarning('no log level found')
+      return
+
+    event_data.log_level = log_level[0]
+
+    user_and_database = self._GetValueFromStructure(
+        structure, 'user_and_database')
+    if user_and_database:
+      event_data.user = ''.join(user_and_database)
+
+    log_line = self._GetValueFromStructure(structure, 'log_line')
+    if log_line:
+      event_data.log_line = log_line.lstrip().rstrip()
+
+    time_zone_string = self._GetValueFromStructure(structure, 'time_zone')
+
+    try:
+      # TODO: PostgreSQL time zone to Python time zone mappings.
```

Comment:
@Fryyyyy I've made some change, here is the issue that is currently breaking the parser on the last log line

Reply:
I've extracted the time zone mappings from PostgreSQL and now attempt to convert them